### ISSUE TITLE
Provide a JvmGenericTypeValidator

### ISSUE THREAD

**Issue by LorenzoBettini** [eclipse/xtext-extras#513]
I'd like to propose an idea I've been thinking of for a while.

When implementing an Xbase language you have to repeat several typical checks, like no duplicate fields, no duplicate methods (according to Java overriding and overloading rules), cyclic hierarchy, etc., like we recently did in https://github.com/eclipse/xtext-eclipse/issues/1205 and https://github.com/eclipse/xtext-eclipse/issues/1205 . I seem to understand that the best way of doing that is by performing checks directly on the inferred JvmModel, like we did in the issues above and like Xtend itself does. After all, it does not make sense to have an inferred JvmModel which then produces incorrect Java code...

So why not providing some kind of `JvmGenericTypeValidator` that does all these checks automatically on the inferred JvmModel and then generating issues the original elements in the program (by means of `IJvmModelAssociations`)? With that respect the starting point could be the `XtendValidator` itself, from where we can extract most of the logic of typical JvmModel validations in the reusable `JvmGenericTypeValidator` (and then `XtendValidator` could be refactored accordingly). The `JvmGenericTypeValidator` could follow the same strategy of `JvmTypeReferencesValidator` which is automatically enabled (by relying on `getEPackages`).

All Xbase languages would then benefit from such automatic validations.

@szarnekow do you think it is feasible, makes sense, could be useful?

**Comment by szarnekow** [comment]
@LorenzoBettini Yes that does make a lot of sense. Having sensible and efficient impl for this and relieving clients from the burden to implement these manually sounds very good.
One caveat: Some Xbase languages may abuse the JVM model to infer class-like structures for other languages with other constraints. There should be an easy way to opt out from the validation.
Also the error messages must be easy to customise since languages often don't want to expose the JVM terminology but the DSL concepts to the end user.

**Comment by LorenzoBettini** [comment]
@szarnekow thank you for the feedback!
I'll try to start working on some prototype soon (hopefully).
Do you have any suggestion to opt out? simple runtimemodule approach?
Moreover, is there an existing test Xbase language that I can use for working on this? I'd need at least the concepts of mapped classes, interface and methods. The Domainmodel could be a good candidate, but I'd rather use some Xbase language that's part of the test suite.

**Comment by szarnekow** [comment]
Seems the only things we have are in org.eclipse.xtext.xbase.testlanguages.
And Xtend of course.

**Comment by szarnekow** [comment]
> Do you have any suggestion to opt out? simple runtimemodule approach?

The validators should be language specific (as in: not registered only for the EPackage but also enabled for the language at hand). See `AbstractInjectableValidator.isLanguageSpecific()`
